### PR TITLE
Fix major bug with MultiGet, DeleteRange, and memtable Bloom

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: MemTable::MultiGet was not considering range tombstones before
querying Bloom filter. This means range tombstones would be skipped for
keys (or prefixes) with no other entries in the memtable. This could cause
old values for a key (in SST files) to still show up until the range tombstone
covering it has been flushed.

This is fixed by essentially disabling the memtable Bloom filter when there
are any range tombstones. (This could be better optimized in the future, but
good enough for now.)

Did some other cleanup/optimization in the same code to (more than) offset
the cost of checking on range tombstones in more cases. There is now
notable improvement when memtable_whole_key_filtering and prefix_extractor
are used together (unusual), and this makes MultiGet closer to the Get
implementation.

Test Plan: new unit test added. Added memtable Bloom to crash test.

Performance testing
--------------------

Build WAL-only DB (recovers to memtable):
```
TEST_TMPDIR=/dev/shm/rocksdb ./db_bench -benchmarks=fillrandom -num=1000000 -write_buffer_size=250000000
```

Query test command, to maximize sensitivity to the changed code:
```
TEST_TMPDIR=/dev/shm/rocksdb ./db_bench -use_existing_db -readonly -benchmarks=multireadrandom -num=10000000 -write_buffer_size=250000000 -memtable_bloom_size_ratio=0.015 -multiread_batched -batch_size=24 -threads=8 -memtable_whole_key_filtering=$MWKF -prefix_size=$PXS
```
(Note -num here is 10x larger for mostly memtable misses)

Before & after run simultaneously, average over 10 iterations per data point, ops/sec.

MWKF=0 PXS=0 (Bloom disabled)
Before: 5724844
After: 6722066

MWKF=0 PXS=7 (prefixes hardly unique; Bloom not useful)
Before: 9981319
After: 10237990

MWKF=0 PXS=8 (prefixes unique; Bloom useful)
Before:  12081715
After: 12117603

MWKF=1 PXS=0 (whole key Bloom useful)
Before: 11944354
After: 12096085

MWKF=1 PXS=7 (whole key Bloom useful in new version; prefixes not useful in old version)
Before: 9444299
After: 11826029

MWKF=1 PXS=7 (whole key Bloom useful in new version; prefixes useful in old version)
Before: 11784465
After: 11778591

Only in this last case is the 'before' *slightly* faster, perhaps because hashing prefixes is slightly faster than hashing whole keys. Otherwise, 'after' is faster.

### CODE REVIEW

File: db/memtable.cc [hunk 1]
```diff
@@ -962,51 +963,64 @@ void MemTable::MultiGet(const ReadOptions& read_options, MultiGetRange* range,
...
+        iter->max_covering_tombstone_seq = std::max(
+            iter->max_covering_tombstone_seq,
+            range_del_iter->MaxCoveringTombstoneSeqnum(iter->lkey->user_key()));
+        // TODO: consider not counting these as Bloom hits to more closely
+        // match bloom_sst_hit_count
+        PERF_COUNTER_ADD(bloom_memtable_hit_count, 1);
+      } else if (whole_key) {
+        bloom_keys[num_keys] = iter->ukey_without_ts;
+        range_indexes[num_keys++] = iter.index();
       } else if (prefix_extractor_->InDomain(iter->ukey_without_ts)) {
```
Comment (ajkr) [inline db/memtable.cc:986]: Was it wrong to do this without checking `memtable_whole_key_filtering`?

**Review by ajkr — APPROVED** [review]
LGTM

File: db/memtable.cc (hunk 1 above)
Comment (anand1976) [reply db/memtable.cc:986]: Don't think so. The comment for memtable_prefix_bloom_size_ratio says the feature is disabled if both prefix_extractor and memtable_whole_key_filtering are not set. And if if both are set but key is not in domain, we would have just skipped the bloom filter and called GetFromTable.

**Review by anand1976 — APPROVED** [review]
LGTM. Great catch!

File: db/memtable.cc (hunk 1 above)
Comment (pdillinger) [reply db/memtable.cc:986]: Like Anand said, just suboptimal, not wrong.

**Comment by pdillinger** [comment]
I am reconsidering the approach based on the apparent small perf regression.

**Comment by pdillinger** [comment]
A subtle quirk of this implementation is that any range tombstones effectively disable the memtable bloom filter. Making that more clear in the code should simplify & optimize a bit.

File: db/memtable.cc [hunk 2]
```diff
@@ -960,53 +963,58 @@ void MemTable::MultiGet(const ReadOptions& read_options, MultiGetRange* range,
...
+  // range tombstones. This is the simplest way to ensure range tombstones are
+  // handled. TODO: allow Bloom checks where max_covering_tombstone_seq==0
+  bool no_range_del = read_options.ignore_range_deletions ||
+                      is_range_del_table_empty_.load(std::memory_order_relaxed);
   MultiGetRange temp_range(*range, range->begin(), range->end());
-  if (bloom_filter_) {
-    std::array<Slice*, MultiGetContext::MAX_BATCH_SIZE> keys;
-    std::array<bool, MultiGetContext::MAX_BATCH_SIZE> may_match = {{true}};
-    autovector<Slice, MultiGetContext::MAX_BATCH_SIZE> prefixes;
+  if (bloom_filter_ && no_range_del) {
```
Comment (ajkr) [inline db/memtable.cc:972]: You want to skip all filter checks when any range tombstone is present? I thought in the previous version of this PR it only skipped checking for a given lookup key when a range tombstone covering that particular lookup key was found.

Comment (ajkr) [reply db/memtable.cc:972]: Never mind, it was always like that. I guess you succeeded at "Making that more clear in the code"

Comment (pdillinger) [reply db/memtable.cc:972]: > You want to skip all filter checks when any range tombstone is present?

I think that's best for now (backport-able with low risk)

> I thought in the previous version ...

Looks can be deceiving. If you dig into details, `range_del_iter == nullptr` is logically equivalent to `no_range_del`. I believe you'd have to check `max_covering_tombstone_seq==0` for coverage check, but I'm not super confident and that's more than I want to write unit tests for at the moment.

Comment (ajkr) [reply db/memtable.cc:972]: Sorry let me start from the beginning. I originally wondered why we skip filter at all when there are range tombstones. They seem useful whether we found a range tombstone or not. Perhaps that can be explained in the comment/TODO -- or if we don't need to skip, perhaps it can give the optimization without introducing edge cases to test.
